# Build Classification Model

In [1]:
import pandas as pd
cuisines_df = pd.read_csv("../data/cleaned_cuisines.csv")
cuisines_df.head()

,cuisine,almond,angelica,anise,anise_seed,apple,apple_brandy,apricot,armagnac,artemisia,...,whiskey,white_bread,white_wine,whole_grain_wheat_flour,wine,wood,yam,yeast,yogurt,zucchini
0,indian,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,indian,1,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,indian,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,indian,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,indian,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1,0


In [2]:
cuisines_label_df = cuisines_df['cuisine']
cuisines_label_df.head()

0    indian
1    indian
2    indian
3    indian
4    indian
Name: cuisine, dtype: object

In [3]:
cuisines_feature_df = cuisines_df.drop(['cuisine'], axis=1)
cuisines_feature_df.head()

,almond,angelica,anise,anise_seed,apple,apple_brandy,apricot,armagnac,artemisia,artichoke,...,whiskey,white_bread,white_wine,whole_grain_wheat_flour,wine,wood,yam,yeast,yogurt,zucchini
0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1,0


In [4]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.svm import LinearSVC
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import accuracy_score,precision_score,confusion_matrix,classification_report, precision_recall_curve
import numpy as np

# Linear SVC classifier

Support-Vector clustering (SVC) is a child of the Support-Vector machines family of ML techniques (learn more about these below). In this method, you can choose a 'kernel' to decide how to cluster the labels. The 'C' parameter refers to 'regularization' which regulates the influence of parameters. The kernel can be one of several; here we set it to 'linear' to ensure that we leverage linear SVC. Probability defaults to 'false'; here we set it to 'true' to gather probability estimates. We set the random state to '0' to shuffle the data to get probabilities.

In [5]:
X_train, X_test, y_train, y_test = train_test_split(cuisines_feature_df, cuisines_label_df, test_size=0.3)

There is a difference in use .svm.SCV(Kernel = 'Linear') and .svm.LinearSVC()

In [9]:
C = 10
# Create different classifiers.
classifiers = {
    'Linear SVC': SVC(kernel='linear', C=C, probability=True,random_state=0),
    'SVC': SVC(),
    'SVC C10': SVC(C=C),
    'svm.LinearSVC':LinearSVC(C = C),
    'KNN classifier': KNeighborsClassifier(C),
    'RFST': RandomForestClassifier(n_estimators=100),
    'ADA': AdaBoostClassifier(n_estimators=100)
}

In [10]:
n_classifiers = len(classifiers)

for index, (name, classifier) in enumerate(classifiers.items()):
    classifier.fit(X_train, np.ravel(y_train))

    y_pred = classifier.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print("Accuracy (train) for %s: %0.1f%% " % (name, accuracy * 100))
    print(classification_report(y_test,y_pred))

Accuracy (train) for Linear SVC: 78.6% 
              precision    recall  f1-score   support

     chinese       0.72      0.72      0.72       250
      indian       0.89      0.88      0.88       238
    japanese       0.72      0.74      0.73       234
      korean       0.85      0.77      0.81       241
        thai       0.76      0.82      0.79       236

    accuracy                           0.79      1199
   macro avg       0.79      0.79      0.79      1199
weighted avg       0.79      0.79      0.79      1199

Accuracy (train) for SVC: 81.7% 
              precision    recall  f1-score   support

     chinese       0.80      0.72      0.76       250
      indian       0.89      0.90      0.90       238
    japanese       0.76      0.81      0.78       234
      korean       0.89      0.80      0.84       241
        thai       0.76      0.86      0.81       236

    accuracy                           0.82      1199
   macro avg       0.82      0.82      0.82      1199
weig

c:\Users\clayton jhordan\.conda\envs\Clayton_Jhordan\lib\site-packages\sklearn\svm\_classes.py:31: FutureWarning: The default value of `dual` will change from `True` to `'auto'` in 1.5. Set the value of `dual` explicitly to suppress the warning.
  warnings.warn(
c:\Users\clayton jhordan\.conda\envs\Clayton_Jhordan\lib\site-packages\sklearn\svm\_base.py:1237: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Accuracy (train) for svm.LinearSVC: 80.8% 
              precision    recall  f1-score   support

     chinese       0.77      0.71      0.74       250
      indian       0.89      0.89      0.89       238
    japanese       0.76      0.79      0.77       234
      korean       0.84      0.81      0.83       241
        thai       0.78      0.85      0.81       236

    accuracy                           0.81      1199
   macro avg       0.81      0.81      0.81      1199
weighted avg       0.81      0.81      0.81      1199

Accuracy (train) for KNN classifier: 74.0% 
              precision    recall  f1-score   support

     chinese       0.71      0.68      0.69       250
      indian       0.86      0.81      0.84       238
    japanese       0.62      0.85      0.72       234
      korean       0.92      0.59      0.72       241
        thai       0.70      0.79      0.74       236

    accuracy                           0.74      1199
   macro avg       0.76      0.74      0.74 

c:\Users\clayton jhordan\.conda\envs\Clayton_Jhordan\lib\site-packages\sklearn\ensemble\_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


Accuracy (train) for ADA: 70.6% 
              precision    recall  f1-score   support

     chinese       0.66      0.57      0.61       250
      indian       0.81      0.85      0.83       238
    japanese       0.67      0.62      0.64       234
      korean       0.71      0.79      0.75       241
        thai       0.67      0.72      0.69       236

    accuracy                           0.71      1199
   macro avg       0.70      0.71      0.70      1199
weighted avg       0.70      0.71      0.70      1199



## Conclusion
1. for the SVM family: LinearSVC: 80.8% | SVC = 81.7% | SVC(kernel = 'linear') = 78.6%
2. KNN show a poor performance.
3. Random Forest appears like the best, AddaBoost show a poor performance.